### PR TITLE
Converting neo to a scaffold generator

### PR DESCRIPTION
**Pull request by eliperelman**

### CODE REVIEW

**Comment by eliperelman** [comment]
To test this out:

- clone and checkout this branch
- `npm link`
- create dummy  and node_modules directory, e.g. `mkdir -p dummy/node_modules && cd dummy`
- `npm link mozilla-neo`
- `node_modules/.bin/neo init`

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "presets": [
+    "es2015",
+    "stage-0",
+    "react"
+  ]
+}
```
Comment (garbas) [inline .babelrc]: why did you move babel configuration out of webpack.config.js?

i see this breaking the library pattern since there is no way (as a user of neo) to only add extend babel configuration on top of neo's babel configuration.

Comment (eliperelman) [inline .babelrc]: There are 2 `.babelrc` files. This one is so we can build the `app.js` file for the consuming project (which was way easier than trying to get the dependent project to build it for itself).

The second `.babelrc` file is in the dependent project at `commands/init/templates/.babelrc`. The project can customize this however they want for their project, the webpack file brings this in.

Comment (eliperelman) [inline .babelrc]: That's to say `babel-loader` will bring in the adjacent `.babelrc` file for the project.

Comment (garbas) [inline .babelrc]: maybe you could read this file at babel loader in config/webpack.config.js ... that we would get best of both.

Comment (eliperelman) [inline .babelrc]: Yes, babel-loader is the one that reads this file.

Comment (eliperelman) [inline .babelrc]: Sorry, not *this* file, but the one at `commands/init/templates/.babelrc`.

File: commands/init/templates/.babelrc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "presets": [
+    "es2015",
+    "stage-0",
+    "react"
+  ]
+}
```
Comment (garbas) [inline commands/init/templates/.babelrc]: i would like to see that this file is not necessary since all babel configuration should come with webpack. then user can create this .babelrc or just use what webpack is doing. am i missing something?

File: commands/init/templates/.eslintrc [hunk 3]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (garbas) [inline commands/init/templates/.eslintrc]: eslint configuration already comes with webpack. why do we need it here?

Comment (eliperelman) [inline commands/init/templates/.eslintrc]: To give a centralized place for anyone to configure their linting rules. Are you saying you would prefer all of this to be in the core webpack configuration and only overrides go in the project `.eslintrc`?

File: commands/init/templates/.babelrc (hunk 2 above)
Comment (eliperelman) [inline commands/init/templates/.babelrc]: Let's say a user doesn't want to use the `stage-0` present and instead use `stage-2`. How would they easily accomplish that? The addition of presets is easy enough to merge, but changes are not.

**Comment by garbas** [comment]
+100 for the neo commands :)

i see it a bit too opinionated if we provide that horrible src/components/<home|nav|layout>.js directory structure. couldn't we just include everything in index.js and let the users decide how/where to put it?

my personal opinion is that less is more and less files that user of neo needs to know about the better. i think we can agree on that. otherwise we're no better then any other react/redux boilerplate example out there. in my perfect world we wouldn't have to generate anything. you would open a file and start writing the application. no worries about how to setup build system ... just straight to coding. Elm is nice example of how this developer UX is done right.

anyway ... there are few things (i left comments inline) which i think we can/should remove from the initial package to keep things really simple ... but overall i'm very happy with the changes :100:

File: commands/init/templates/.travis.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+language: node_js
+node_js:
+  - "6"
+script: npm test
+before_install:
+  - export CHROME_BIN=chromium-browser
```
Comment (garbas) [inline commands/init/templates/.travis.yml]: maybe we should use firefox :)

Comment (eliperelman) [inline commands/init/templates/.travis.yml]: 😄 Certainly, just have to figure out how to do it.

File: commands/init/templates/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# <%= data.appTitle %>
```
Comment (garbas) [inline commands/init/templates/README.md]: could we include something like [this](http://docs.writethedocs.org/writing/beginners-guide-to-docs/#id1) in the generated readme?

Comment (garbas) [inline commands/init/templates/README.md]: maybe also a travis badge

**Comment by eliperelman** [comment]
@garbas thanks, I definitely feel like we are moving in the right direction, even if it's not perfect yet.

Concerning the components directory, it is only organized that way to keep the components separate from the `index` and `routes`. Neo commands have no knowledge of the `src` internal structure, only the tests do from referencing the components. There's nothing stopping it from being re-organized within the project.

File: commands/init/templates/karma.conf.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+const config = require('mozilla-neo/config/karma');
+
+module.exports = (karma) => karma.set(config);
```
Comment (garbas) [inline commands/init/templates/karma.conf.js]: could we should already how to extend above config, like it was before ... with empty  ``{ // your custom config comes here }``

File: commands/init/templates/README.md (hunk 5 above)
Comment (eliperelman) [inline commands/init/templates/README.md]: 👍 on the travis badge, but I'm confused about the readme. What do you want added?

File: commands/init/templates/package.json [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  "dependencies": {
+    "babel-preset-es2015": "6.6.0",
+    "babel-preset-react": "6.5.0",
+    "babel-preset-stage-0": "6.5.0",
+    "bootstrap": "4.0.0-alpha.2",
+    "expose-loader": "0.7.1",
+    "ghooks": "1.2.1",
+    "html-webpack-plugin": "2.16.1",
+    "jquery": "2.2.3",
+    "mozilla-neo": "1.0.0",
```
Comment (garbas) [inline commands/init/templates/package.json]: i was under impression that only ``mozilla-neo`` would be a dependency. is this not the case? it is working for me like this (with node v6)

File: commands/init/templates/karma.conf.js (hunk 6 above)
Comment (eliperelman) [inline commands/init/templates/karma.conf.js]: Yes, this is something I was thinking about too. Right now I can't think of use cases of extending the test config, but extension should be pretty easy:

```js
const config = require('mozilla-neo/config/karma');
const custom = require('./custom.karma');

module.exports = (karma) => karma.set(deepmerge(config, custom));
```

And the only reason that this logic is pushed into the project is because karma has a **reeaaaaaly* hard time working with things where the config isn't co-located with current working directory.

Comment (eliperelman) [inline commands/init/templates/karma.conf.js]: In fact, I'll make that change (inline).

File: commands/init/templates/src/components/layout/index.scss [hunk 8]
```diff
@@ -1,2 +1 @@
-
 @import "~bootstrap/scss/bootstrap";
```
Comment (garbas) [inline commands/init/templates/src/components/layout/index.scss]: maybe we should use [mozmaker](https://github.com/mozilla/mozmaker) (a bootstrap v4 with firefox colors and fonts)

Comment (eliperelman) [inline commands/init/templates/src/components/layout/index.scss]: I don't want to get opinionated about theme since Tools will have a custom bootstrap theme. Not to mention it won't be using `bootstrap` but `bootstrap-flex`.

File: commands/init/templates/package.json (hunk 7 above)
Comment (eliperelman) [inline commands/init/templates/package.json]: There are some dependencies that belong to the project's front-end and I want this customizable. Like `jquery`, most likely tools will not be using it. `expose` is only needed because of `jquery`. `html-webpack-plugin` needs project configuration settings and so this belongs to the project. `ghooks` belongs to the project. The babel presets are debatable.

File: commands/init/templates/webpack.config.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    new HtmlPlugin({
+      template: path.join(__dirname, 'src/template.ejs'),
+      hash: true,
+      xhtml: true,
+      title: '<%= data.appTitle %>',
+      description: '<%= data.description %>',
+      author: '<%= data.author %>'
+    })
+  ]
+};
```
Comment (garbas) [inline commands/init/templates/webpack.config.js]: you dont really know that you are extending anything and ppl will get confused by it

i would prefer if we ``require('mozilla-neo/config/webpack')`` and extend it (using ``webpack-merge``) and also in comments provide a link to what we are extending and link to possible configuration options (to webpack docs page)

File: commands/init/templates/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+const HtmlPlugin = require('html-webpack-plugin');
+const path = require('path');
+
+module.exports = {
+  plugins: [
+    new HtmlPlugin({
```
Comment (garbas) [inline commands/init/templates/webpack.config.js]: why is HtmlPlugin not already part of ``mozilla-neo``?

File: commands/init/templates/webpack.config.js (hunk 9 above)
Comment (eliperelman) [inline commands/init/templates/webpack.config.js]: I think I can see doing the merge here.

File: commands/init/templates/webpack.config.js (hunk 10 above)
Comment (eliperelman) [inline commands/init/templates/webpack.config.js]: The data consumed by this plugin is generated from running `neo init`. If the project needs its own dependency, it should be explicit about it.

Comment (eliperelman) [inline commands/init/templates/webpack.config.js]: I can be convinced to change this, but I'm not so sure yet, so I'd like to hear if there is a better solution.

File: .babelrc (hunk 1 above)
Comment (garbas) [inline .babelrc]: but this still doesn't solve the problem i have. where i want mozilla-neo to provide me with the default set of babel presents. and i don't want to worry about it. as a user of mozilla-neo i don't even want to know what a babel preset is.

File: commands/init/templates/.babelrc (hunk 2 above)
Comment (garbas) [inline commands/init/templates/.babelrc]: they would take ``mozilla/config/webpack`` which is just an object and provide new list or extend it. in 99% cases they wont even touch it.

File: commands/init/templates/.eslintrc (hunk 3 above)
Comment (garbas) [inline commands/init/templates/.eslintrc]: yes, i see an average user doesn't want to pick this things. they just want to start writing code. we select the eslint style for then ... but yes _if_ and when they want to change the style they can create ``.eslintrc`` ... but not by default.

Comment (eliperelman) [inline commands/init/templates/.eslintrc]: Hmm ok, let me think about the implications about this.

File: commands/init/templates/README.md (hunk 5 above)
Comment (garbas) [inline commands/init/templates/README.md]: you see the readme template down bellow? i kinda guides ppl to write better readme's ... i use it all the times

File: commands/init/templates/karma.conf.js (hunk 6 above)
Comment (garbas) [inline commands/init/templates/karma.conf.js]: what about just in place

```
const config = require('mozilla-neo/config/karma');
const custom = require('./custom.karma');

module.exports = (karma) => karma.set(deepmerge(config, {
  // your custom config goes here
  // read about the default here -> https://...
  // and more about karma configuration options you can find here -> https:///
}));
```

File: commands/init/templates/webpack.config.js (hunk 10 above)
Comment (garbas) [inline commands/init/templates/webpack.config.js]: basically i dont want to worry with any html. that index.html should just be there. most of the times you will just provide the title and that's it, no?

File: commands/init/templates/package.json (hunk 7 above)
Comment (garbas) [inline commands/init/templates/package.json]: +1 undestand now ... i guess only babel presets (if we integrate babelrc in webpack) can be removed

File: commands/init/templates/src/components/layout/index.scss (hunk 8 above)
Comment (garbas) [inline commands/init/templates/src/components/layout/index.scss]: just a suggestions ... since it's a mozilla project maybe a mozilla colors/fonts would be a nice default.

**Comment by garbas** [comment]
@eliperelman awesome ... getting excited about merging this PR

i think i was to vague about who (by my opinion) target users of mozilla-neo are. from relengapi perspective those are more or less python devs or ops with some knowledge how to script jquery stuff. all those build tools webpack/babel/karma/... are foreign to them and they don't care about it and prefer to have tools and configuration chosen for them (and also maintained). they just want to write code (like in jquery days).

ofcourse pleasing the above persona must not prevent an advance user to be able to configure tools, we chose, by their liking.

i would just like to see defaults be more targeted towards less advanced users.

File: commands/init/templates/src/components/layout/index.scss (hunk 8 above)
Comment (garbas) [inline commands/init/templates/src/components/layout/index.scss]: isn't bootstrap v4 already supporting flex?

File: commands/init/templates/webpack.config.js (hunk 10 above)
Comment (eliperelman) [inline commands/init/templates/webpack.config.js]: Remember that all this stuff is already configured for them by running `neo init`. They don't need to configure webpack, karma, eslint, babel, the HTML plugin, or anything to get started. They can literally run `neo init` then `npm run build` and have the site built.

File: commands/init/templates/src/components/layout/index.scss (hunk 8 above)
Comment (eliperelman) [inline commands/init/templates/src/components/layout/index.scss]: Yes, it is, but I don't want to force that opinion on anyone.

Comment (eliperelman) [inline commands/init/templates/src/components/layout/index.scss]: If you want to make the mozmaker theme the default, let's get this PR finished up and we can add that in a different PR.

File: commands/init/templates/karma.conf.js (hunk 6 above)
Comment (eliperelman) [inline commands/init/templates/karma.conf.js]: Yes exactly, that's what I meant about inline, sorry. Except that the configuration will be in a variable, so it is explicit.

File: .babelrc (hunk 1 above)
Comment (eliperelman) [inline .babelrc]: Neo **does** provide you with a default set of babel presets, and loads them for you, it just does this by using its own file as opposed to configuring it in the `webpack.config.js` file. A user does not have to know what this file is for or how to use it in order to have it work, it's loaded with the project on init.

File: commands/init/templates/.babelrc (hunk 2 above)
Comment (eliperelman) [inline commands/init/templates/.babelrc]: While this may fit many use cases, our merge will not let you easily change the defaults, rather only add to it. Moving the configuration solves all use cases, whether that's in it's own file or located in the `webpack.config.js` file.

File: commands/init/templates/.eslintrc (hunk 3 above)
Comment (eliperelman) [inline commands/init/templates/.eslintrc]: The `.eslintrc` file is created automatically on init; there is nothing that needs to be picked. The file exists to configure the lint rules, and most users won't need to mess with this at all. They only need to modify the file if they want to change lint rules.

The existence of the `webpack.config.js` and `karma.conf.js` and `.babelrc` and `.eslintrc` in the project as opposed to the parent is for the same reason the `.travis.yml` file lives in the project and is not modified unless changes are wanted. We aren't importing the Travis configuration from Neo, and yet it's OK for this file to live in the project.

File: commands/init/templates/webpack.config.js (hunk 10 above)
Comment (garbas) [inline commands/init/templates/webpack.config.js]: ``neo init`` should generate as little as code as possible. preferable none if possible. i'm not worried that you have to type this yourself. ``neo init`` you will do once a project, but following best practices with neo after that is something we should excel at.

if we can remove something from ``neo init`` and make it **just work** then i think we should.

File: .babelrc (hunk 1 above)
Comment (garbas) [inline .babelrc]: if you're talking about generating a file with ``neo init`` then yes i agree ... but is not coming from neo. at one point in time a person generated this file. and if best practices change everybody will have to update manually. i would like to avoid that.

File: commands/init/templates/.eslintrc (hunk 3 above)
Comment (garbas) [inline commands/init/templates/.eslintrc]: i think (as commented on other place) we see things differently:
 - you would like to generate them via ``neo init``
 - i would like them to be provided/imported from neo 

i think if we make them importable from neo also your option of having them configurable via those files is still possible. or am i missing something.

if possible i would love to also see  other dotfiles being removed from the ``neo init``, but i don't see a way how to do it (yet :P). please read this as you can configure via dot files, but you don't have to if default neo condifuration works for you.

File: commands/init/templates/README.md (hunk 5 above)
Comment (eliperelman) [inline commands/init/templates/README.md]: README updated, but configuring a travis badge can't be automated unless we know the repo information.

**Comment by eliperelman** [comment]
@garbas I think there are a couple more changes, but I think we are much closer to what you want now.

There is very little configuration in the project now, but it should be easily changeable, which was my concern. Now after running `init`, if you want to change the webpack config, you can create a file and override it in the command, which seems conventional:

```sh
npm i -D webpack-merge
```

```js
// In $PROJECT_ROOT/webpack.dev.js
const config = require('mozilla-neo/config/webpack.dev');
const merge = require('webpack-merge');

module.exports = merge(config, {
  devtool 'cheap-inline-source-map'
});
```

```json
{
  "scripts": {
    "start": "neo start --port 4000 --config webpack.dev.js"
  }
}
```

It's not perfect, but I think this is getting pretty close. :smile:

**Comment by eliperelman** [comment]
FYI I still need to update the project README to document how to override the configs.

File: commands/test/index.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+const path = require('path');
+const { Server } = require('karma');
+
+module.exports = (args) => {
+  new Server({
+    configFile: args.options.file ?
+      path.resolve(process.cwd(), args.options.file) :
```
Comment (garbas) [inline commands/test/index.js]: i love the idea of getting away with ``webpack.config.js`` ... i guess we could also do it for ``karma.conf.js``?

**Comment by garbas** [comment]
@eliperelman you even removed ``webpack.config.js`` and ``karma.conf.js`` from initial template. awesome, just awesome!! :100:

which means whole npm init is few files! i love it. really great work.
i say merge and lets fix the README as we go.

File: commands/test/index.js (hunk 11 above)
Comment (eliperelman) [inline commands/test/index.js]: I'm not following. This allows us to get away from webpack and karma configurations in the project altogether unless wanted.

**Comment by eliperelman** [comment]
@garbas haha thanks. Removing `karma.conf.js` was the hardest. I'll get this fixed up tomorrow.

File: commands/test/index.js (hunk 11 above)
Comment (garbas) [inline commands/test/index.js]: sorry my bad ... i misread the code.

**Comment by garbas** [comment]
:tada:
